agent/agent.go

```diff
@@ -0,0 +1,825 @@
+	wd := o.opts.WorkingDirectory()
+	o.logger.Infof("removing contents from working directory: %s", wd)
+	if err := fs.RemoveContents(wd); err != nil {
+		return nil, grpc.Errorf(codes.Internal, "unable to clear working directory: %v", err)
+	}
+
+	// initialize any resources needed on the host
+	o.logger.Infof("initializing host resources")
+	if err := o.opts.InitHostResourcesFn()(); err != nil {
+		o.resetWithLock(errSetupInitializeHostResources) // release any resources
+		return nil, grpc.Errorf(codes.Internal, "unable to initialize host resources: %v", err)
+	}
+
+	// setup new heartbeating
+	if request.HeartbeatEnabled {
+		o.heartbeatTimeoutCh = make(chan struct{})
+		beater, err := newHeartbeater(o, o.heartbeatTimeoutCh, request.OperatorUuid, request.HeartbeatEndpoint, o.opts.InstrumentOptions())
+		if err != nil {
+			o.resetWithLock(errSetupInitializeHostResources) // release any resources
+			return nil, grpc.Errorf(codes.Aborted, "unable to start heartbeating process: %v", err)
+		}
+		o.heartbeater = beater
+		go o.monitorHeartbeatTimeout(o.heartbeatTimeoutCh)
+		o.heartbeater.start(time.Second * time.Duration(request.HeartbeatFrequencySecs))
+	}
+
+	o.token = request.SessionToken
+	return &m3em.SetupResponse{}, nil
+}
+
+func (o *opAgent) monitorHeartbeatTimeout(timeoutCh chan struct{}) {
+	for range timeoutCh {
+		o.logger.Warnf("heartbeat sending timed out, resetting agent")
+		o.Lock()
+		err := o.resetWithLock(nil) // nil indicates we don't want to send a heartbeat
+		o.Unlock()
+		if err == nil {
+			o.logger.Infof("successfully reset agent")
+		} else {
+			o.logger.Warnf("error while resetting agent: %v", err)
+		}
+		// resetting breaks out of the loop as it closes the timeoutCh
+	}
+}
+
+// non-thread safe multi-writer
+type multiWriter struct {
+	io.Closer
+	fds []*os.File
+}
+
+// based on signature of os.OpenFile
+func newMulitWriter(paths []string, flags int, fileMode os.FileMode, dirMode os.FileMode) (*multiWriter, error) {
+	mw := &multiWriter{}
+	for _, p := range paths {
+		// ensure directory exists
+		dir := filepath.Dir(p)
+		err := os.MkdirAll(dir, dirMode)
+		if err != nil {
+			mw.Close() // cleanup
+			return nil, err
+		}
+
+		// create file
+		fd, err := os.OpenFile(p, flags, fileMode)
+		if err != nil {
+			mw.Close() // cleanup
+			return nil, err
+		}
+		mw.fds = append(mw.fds, fd)
+	}
+	return mw, nil
+}
+
+// mimics signature of os.File.Write
+func (mw *multiWriter) write(b []byte) (int, error) {
+	var (
+		first = true
+		n     int
+		err   error
+	)
+
+	// write the provided data to all fds, terminating
+	// early if any fd write fails or is not the same
+	// as the others
+	for _, fd := range mw.fds {
+		// initialize for the first fd
+		if first {
+			n, err = fd.Write(b)
+			first = false
+			if err != nil {
+				return n, err
+			}
+			continue
+		}
+
+		// for the rest of the fds, ensure
+		// they all write the same number of bytes
+		// as the first one
```

Comment:
Fair enough, according to the [docs]( https://golang.org/pkg/os/#File.Write) checking only the error is sufficient.